workspaces/diff-engine/cli/src/main.rs

```diff
@@ -55,115 +83,170 @@ fn main() {
-                  // TODO: Find way to actually write error info
+    }
+  });
+
+  tokio::pin!(results_manager);
+
+  dbg!("waiting for next interaction");
+
+  let diffing_interactions = async move {
+    let diff_results = interaction_lines
+      .map(Ok)
+      .try_for_each_concurrent(diff_queue_size, |interaction_json_result| {
+        let projection = spec_projection.clone();
+        let results_sender = results_sender.clone();
+
+        let diff_task = tokio::spawn(async move {
+          let diff_comp = tokio::task::spawn_blocking::<
+            _,
+            Option<(Vec<InteractionDiffResult>, Tags)>,
+          >(move || {
+            let interaction_json =
+              interaction_json_result.expect("can read interaction json line from stdin");
+            let TaggedInput(interaction, tags): TaggedInput<HttpInteraction> =
+              match serde_json::from_str(&interaction_json) {
+                Ok(tagged_interaction) => tagged_interaction,
+                Err(parse_error) => {
+                  eprintln!("could not parse interaction json: {}", parse_error);
+                  return None;
                 }
+              };
+
+            Some((diff_interaction(&projection, interaction), tags))
+          });
+          //dbg!("waiting for results");
+          let results = diff_comp
+            .await
+            .expect("diffing of interaction should be successful");
+          //dbg!("got results");
+
+          if let Some((results, tags)) = results {
+            for result in results {
+              //dbg!(&result);
+              if let Err(_) = results_sender
+                .send(ResultContainer::from((result, &tags)))
+                .await
+              {
+                panic!("could not write diff result to results channel");
+                // TODO: Find way to actually write error info
               }
             }
-          });
+          }
+        });
 
-          diff_task
-        })
-        .await;
+        diff_task
+      })
+      .await;
 
-      dbg!("interactions stream closed");
+    dbg!("interactions stream closed");
 
-      drop(results_sender);
-      diff_results
-    };
+    drop(results_sender);
+    diff_results
+  };
+
+  try_join!(diffing_interactions, results_manager).expect("essential worker task panicked");
+}
+
+async fn assemble(spec_events: Vec<SpecEvent>) {
+  let stdout = stdout();
+
+  let mut results_sink = streams::spec_events::into_json_array_items(stdout);
+  results_sink
+    .get_mut()
+    .write_u8(b'[')
+    .await
+    .expect("could not write array start to stdout");
+
+  for spec_event in spec_events {
+    if let Err(_) = results_sink.send(spec_event).await {
+      panic!("could not stream event result to stdout"); // TODO: Find way to actually write error info
+    }
+  }
+  results_sink
+    .get_mut()
+    .write_u8(b']')
+    .await
+    .expect("could not write array start to stdout");
```

Comment:
Writing valid JSON instead of JSONL this way was chosen to make implementation and backwards-compatibility easier. Given that downstream it's never really consumed partially, the main benefit of JSONL is lost anyway. That said, there's no reason why it couldn't support both through use of a flag, so we can move to JSONL in the future.

That would give us an opportunity to move some of this direct writing to stdout to something like `JsonArrayItemSink`, something I deemed outside of the scope of this PR (plenty of time was spent on writing these results as is).